contracts/contracts/strategies/ThreePoolStrategy.sol

```diff
@@ -0,0 +1,256 @@
+     * @param _amount Amount of asset to withdraw
+     * @return amountWithdrawn Amount of asset that was withdrawn
+     */
+    function withdraw(
+        address _recipient,
+        address _asset,
+        uint256 _amount
+    ) external onlyVault returns (uint256 amountWithdrawn) {
+        require(_recipient != address(0), "Invalid recipient");
+        require(_amount > 0, "Invalid amount");
+        // Calculate how much of the pool token we need to withdraw
+        (
+            uint256 contractPTokens,
+            uint256 gaugePTokens,
+            uint256 totalPTokens
+        ) = _getTotalPTokens();
+        // Calculate the max amount of the asset we'd get if we withdrew all the
+        // platform tokens
+        ICurvePool curvePool = ICurvePool(platformAddress);
+        uint256 maxAmount = curvePool.calc_withdraw_one_coin(
+            totalPTokens,
+            poolCoinIndex
+        );
+        // Calculate how many platform tokens we need to withdraw the asset amount
+        uint256 withdrawPTokens = totalPTokens.mul(_amount).div(maxAmount);
+        if (contractPTokens < withdrawPTokens) {
+            // Not enough of pool token exists on this contract, must be staked
+            // in Gauge, unstake
+            ICurveGauge(crvGaugeAddress).withdraw(withdrawPTokens);
+        }
+        curvePool.remove_liquidity_one_coin(withdrawPTokens, poolCoinIndex, 0);
+        IERC20(_asset).transfer(_recipient, _amount);
+        // Transfer any leftover dust back to the vault buffer.
+        uint256 dust = IERC20(_asset).balanceOf(address(this));
+        if (dust > 0) {
+            IERC20(_asset).safeTransfer(vaultAddress, dust);
+        }
+        emit Withdrawal(
+            _asset,
+            address(assetToPToken[_asset]),
+            amountWithdrawn
+        );
+    }
+
+    /**
+     * @dev Remove all assets from platform and send them to Vault contract.
+     */
+    function liquidate() external onlyVaultOrGovernor {
+        // Remove entire balance, 3pool strategies only support a single asset
+        // so safe to use assetsMapped[0]
+        IERC20 asset = IERC20(assetsMapped[0]);
+        uint256 pTokenBalance = IERC20(assetToPToken[address(asset)]).balanceOf(
+            address(this)
+        );
+        ICurvePool(platformAddress).remove_liquidity_one_coin(
+            pTokenBalance,
+            poolCoinIndex,
+            0
+        );
+        // Transfer the asset out to Vault
+        asset.safeTransfer(vaultAddress, asset.balanceOf(address(this)));
+    }
+
+    /**
+     * @dev Get the total asset value held in the platform
+     *  This includes any interest that was generated since depositing
+     *  We calculate this by calculating a what we would get if we liquidated
+     *  the allocated percentage of this asset.
+     * @param _asset      Address of the asset
+     * @return balance    Total value of the asset in the platform
+     */
+    function checkBalance(address _asset)
```

Comment:
This should probably return a zero if the _asset passed along does not match the strategy's asset. That's really going to confuse someone someday.